src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -0,0 +1,1499 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// The exported function names need to be unique (can't be disambiguated based on signature), hence
+// we introduce suffix letters to indicate the general patterns used.
+// * A suffix means aligned and padded for SSE operations.
+// * U suffix means unaligned and unpadded.
+// * P suffix means sparse (unaligned) partial vector - the vector is only part of a larger sparse vector.
+// * Tran means the matrix is transposed.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    internal static class AvxIntrinsics
+    {
+        private static readonly Vector256<float> _absMask256 = Avx.StaticCast<int, float>(Avx.SetAllVector256(0x7FFFFFFF));
+
+        private const int Vector256Alignment = 32;
+
+        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
+        private static bool HasCompatibleAlignment(AlignedArray alignedArray)
+        {
+            Contracts.AssertValue(alignedArray);
+            Contracts.Assert(alignedArray.Size > 0);
+            return (alignedArray.CbAlign % Vector256Alignment) == 0;
+        }
+
+        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
+        private static unsafe float* GetAlignedBase(AlignedArray alignedArray, float* unalignedBase)
+        {
+            Contracts.AssertValue(alignedArray);
+            float* alignedBase = unalignedBase + alignedArray.GetBase((long)unalignedBase);
+            Contracts.Assert(((long)alignedBase & (Vector256Alignment - 1)) == 0);
+            return alignedBase;
+        }
+
+        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<float> GetLow(in Vector256<float> x)
```

Comment:
nit: for these, where the two method signatures are effectively identical, it is probably better to just call `Avx.GetLowerHalf` directly, rather than having an intermediate `GetLow` signature.